### PR TITLE
Automatic insertion of colons after class, def, if...

### PR DESCRIPTION
**Pull request by kelsa-pi**
New PR, replace for (#547)
I am sorry for the inconvenience.

### CODE REVIEW

File: pyzo/codeeditor/extensions/behaviour.py [hunk 1]
```diff
@@ -602,3 +602,27 @@ def _moveCursorRight(self, n):
...
+
+
+class AutoInsertColons(object):
+    """
+    Automatic insertion of colons after class, def, if...
+    """
+
+    def keyPressEvent(self, event):
+
+        if event.key() in (Qt.Key_Enter, Qt.Key_Return):
```
Comment (almarklein) [inline pyzo/codeeditor/extensions/behaviour.py]: This is Python specific, so we should add: `if 'python' in self.parser().name().lower():`

**Comment by almarklein** [comment]
> I am sorry for the inconvenience.
No problem!!


This looks good, and I expect it would work just fine in 95% of cases. But in order to make it work correctly in the remaining 5% this needs (a lot?) more work .... imagine for example:

```py
if (something == 42 and <enter because I want the rest of the test on a new line>

if (something == 42 and
    somethingelse == 7 and
    thatotherthing != 3) <enter, NOW we must insert a colon!>

def foo():  # we should rename this <enter, we need to consider comments and whitespace>

```

The best way to do this is to run through the tokens, see https://github.com/pyzo/pyzo/blob/master/pyzo/codeeditor/extensions/behaviour.py#L454 for an example. The second case is especially tricky; we may need to look at previous lines, possibly more than one. But how would we know that the current line is part of an if/def/etc.?

Doing this correctly requires higher level information then we have now. It might be that if source code parsing (as done by the source structure tool) is combined with the code-editor (#235), this information *is* available. Until then, I think this feature will be very hard to get right, unfortunately.

**Comment by kelsa-pi** [comment]
OK, I'm working on it

**Comment by ysalmon** [comment]
If this gets merged, please do add an option to turn it off like for parentheses autocompletion, as this is something that I surely would not want with students (this would hinder their learning of the syntax).

**Comment by almarklein** [comment]
Closing this. I suspect that an option like this would easily become annoying rather than helpful :)

@kelsa-pi You're free to try and prove me wrong :D  but it's probably good to know I'd be reluctant before putting a lot of enery into this :)

**Pull request closed without merging**